pkg/loop/internal/relayer/pluginprovider/contractreader/contract_reader.go

```diff
@@ -758,6 +856,43 @@ func convertSequencesToVersionedBytesProto(sequences []types.Sequence, version E
 	return pbSequences, nil
 }
 
+func convertSequencesWithKeyToVersionedBytesProto(sequences []types.SequenceWithKey, filters []*pb.ContractKeyFilter, encodeWith EncodingVersion) ([]*pb.SequenceWithKey, error) {
+
+	keyToEncodingVersion := make(map[string]EncodingVersion)
+	for _, filter := range filters {
+		if filter.AsValueType {
+			keyToEncodingVersion[filter.Filter.Key] = ValuesEncodingVersion
+		} else {
+			keyToEncodingVersion[filter.Filter.Key] = encodeWith
+		}
+	}
+
+	var pbSequences []*pb.SequenceWithKey
+	for _, sequence := range sequences {
+		version, ok := keyToEncodingVersion[sequence.Key]
+		if !ok {
+			return nil, fmt.Errorf("missing encoding version for key %s", sequence.Key)
+		}
+
+		versionedSequenceDataType, err := EncodeVersionedBytes(sequence.Data, version)
+		if err != nil {
+			return nil, err
+		}
+		pbSequence := &pb.SequenceWithKey{
+			Key:            sequence.Key,
+			SequenceCursor: sequence.Cursor,
+			Head: &pb.Head{
+				Height:    sequence.Height,
+				Hash:      sequence.Hash,
+				Timestamp: sequence.Timestamp,
```

Comment:
Passing through the `iter.Seq[string,Sequence]` makes this a little more direct, besides the efficiency:
```suggestion
	for key, sequence := range sequences {
		version, ok := keyToEncodingVersion[key]
		if !ok {
			return nil, fmt.Errorf("missing encoding version for key %s", key)
		}

		versionedSequenceDataType, err := EncodeVersionedBytes(sequence.Data, version)
		if err != nil {
			return nil, err
		}
		pbSequence := &pb.SequenceWithKey{
			Key:            sequence.Key,
			SequenceCursor: sequence.Cursor,
			Head: &pb.Head{
				Height:    sequence.Height,
				Hash:      sequence.Hash,
				Timestamp: sequence.Timestamp,
```